In [3]:
install.packages("tidyverse")

also installing the dependencies 'colorspace', 'sys', 'bit', 'ps', 'sass', 'cachem', 'farver', 'labeling', 'munsell', 'RColorBrewer', 'viridisLite', 'rappdirs', 'rematch', 'askpass', 'bit64', 'prettyunits', 'processx', 'highr', 'xfun', 'yaml', 'bslib', 'fontawesome', 'jquerylib', 'tinytex', 'backports', 'generics', 'memoise', 'blob', 'DBI', 'R6', 'tidyselect', 'withr', 'data.table', 'gtable', 'isoband', 'scales', 'gargle', 'cellranger', 'curl', 'ids', 'rematch2', 'cpp11', 'pkgconfig', 'mime', 'openssl', 'timechange', 'systemfonts', 'textshaping', 'clipr', 'vroom', 'tzdb', 'progress', 'callr', 'fs', 'knitr', 'rmarkdown', 'selectr', 'stringi', 'broom', 'conflicted', 'dbplyr', 'dplyr', 'dtplyr', 'forcats', 'ggplot2', 'googledrive', 'googlesheets4', 'haven', 'hms', 'httr', 'lubridate', 'magrittr', 'modelr', 'purrr', 'ragg', 'readr', 'readxl', 'reprex', 'rstudioapi', 'rvest', 'stringr', 'tibble', 'tidyr', 'xml2'




package 'colorspace' successfully unpacked and MD5 sums checked
package 'sys' successfully unpacked and MD5 sums checked
package 'bit' successfully unpacked and MD5 sums checked
package 'ps' successfully unpacked and MD5 sums checked
package 'sass' successfully unpacked and MD5 sums checked
package 'cachem' successfully unpacked and MD5 sums checked
package 'farver' successfully unpacked and MD5 sums checked
package 'labeling' successfully unpacked and MD5 sums checked
package 'munsell' successfully unpacked and MD5 sums checked
package 'RColorBrewer' successfully unpacked and MD5 sums checked
package 'viridisLite' successfully unpacked and MD5 sums checked
package 'rappdirs' successfully unpacked and MD5 sums checked
package 'rematch' successfully unpacked and MD5 sums checked
package 'askpass' successfully unpacked and MD5 sums checked
package 'bit64' successfully unpacked and MD5 sums checked
package 'prettyunits' successfully unpacked and MD5 sums checked
package 'processx' success

In [4]:
# Importing necessary libraries
library("tidyverse")
library("ggplot2")
library("ggrepel")
library("ggcorrplot")
library("DT")

── Attaching core tidyverse packages ──────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


ERROR: Error in library("ggrepel"): there is no package called 'ggrepel'


# Dataset : COVID-19_cases_TX (updated)

In [5]:
dataset_cases_TX <- read.csv("https://raw.githubusercontent.com/yamphurai/Project2_ClusterAnalysis/refs/heads/main/updated_COVID-19_cases_TX.csv")

In [7]:
# First 10 rows
head(dataset_cases_TX)

county_fips_code county_name     date       confirmed_cases deaths
1 48001            Anderson County 2020-01-22 0               0     
2 48001            Anderson County 2020-01-23 0               0     
3 48001            Anderson County 2020-01-24 0               0     
4 48001            Anderson County 2020-01-25 0               0     
5 48001            Anderson County 2020-01-26 0               0     
6 48001            Anderson County 2020-01-27 0               0

In [8]:
# internal structure of the dataset
str (dataset_cases_TX)

'data.frame':	93980 obs. of  5 variables:
 $ county_fips_code: int  48001 48001 48001 48001 48001 48001 48001 48001 48001 48001 ...
 $ county_name     : chr  "Anderson County" "Anderson County" "Anderson County" "Anderson County" ...
 $ date            : chr  "2020-01-22" "2020-01-23" "2020-01-24" "2020-01-25" ...
 $ confirmed_cases : int  0 0 0 0 0 0 0 0 0 0 ...
 $ deaths          : int  0 0 0 0 0 0 0 0 0 0 ...


In [10]:
# Basic statistics
summary(dataset_cases_TX)

 county_fips_code county_name            date           confirmed_cases 
 Min.   :48001    Length:93980       Length:93980       Min.   :     0  
 1st Qu.:48127    Class :character   Class :character   1st Qu.:     2  
 Median :48254    Mode  :character   Mode  :character   Median :    84  
 Mean   :48254                                          Mean   :  2167  
 3rd Qu.:48381                                          3rd Qu.:   645  
 Max.   :48507                                          Max.   :297629  
     deaths       
 Min.   :   0.00  
 1st Qu.:   0.00  
 Median :   2.00  
 Mean   :  38.33  
 3rd Qu.:  15.00  
 Max.   :4024.00  

In [11]:
# Checking for missing values
sum(is.na(dataset_cases_TX))

[1] 0

In [14]:
# count of counties
n_distinct(dataset_cases_TX$county_name)

# count of county FIPs codes
n_distinct(dataset_cases_TX$county_fips_code)

[1] 254

[1] 254

There are 254 different counties with unique code in the dataset

# Dataset : COVID-19_cases_plus_census (updated)

In [16]:
dataset_plus_census <- read.csv("https://raw.githubusercontent.com/yamphurai/Project2_ClusterAnalysis/refs/heads/main/updated_COVID-19_cases_plus_census.csv")

In [17]:
# First 10 rows
head(dataset_plus_census)

state state_fips_code county_name       confirmed_cases deaths total_pop
1 VT    50              Essex County        111             0      6203   
2 VT    50              Chittenden County  3636            78    160985   
3 DE    10              Kent County       11548           187    173145   
4 RI    44              Washington County  5521           122    126190   
5 NH    33              Belknap County     2496            79     60383   
6 RI    44              Newport County     3578             6     83204   
  male_pop female_pop white_pop black_pop ⋯ male_21_49 male_50_above
1  3135     3068        5929       64     ⋯   900       1701        
2 78928    82057      143657     4091     ⋯ 32947      27110        
3 83544    89601      108627    41729     ⋯ 31121      29975        
4 61154    65036      115206     1621     ⋯ 20438      27238        
5 29705    30678       57523      285     ⋯  9663      14520        
6 40952    42252       71549     2596     ⋯ 15262      18187        
  female_0_20 female_21_49 female_50_above unemployed_pop employed_pop commute
1   603         924         1657            149            2784          4820 
2 20668       32285        30864           3827           90054        148357 
3 23801       33109        35761           5563           78078        145421 
4 16106       20580        30387           4142           64576        114755 
5  6541        9882        15317           1589           30674         55288 
6  9040       14232        20449           2349           41471         73711 
  worked_at_home walked_to_work
1  167             81          
2 4783           6859          
3 3270           1379          
4 3110           1725          
5 1487            449          
6 2711           2749

In [18]:
# internal structure of the dataset
str (dataset_plus_census)

'data.frame':	3142 obs. of  33 variables:
 $ state                : chr  "VT" "VT" "DE" "RI" ...
 $ state_fips_code      : int  50 50 10 44 33 44 50 9 50 50 ...
 $ county_name          : chr  "Essex County" "Chittenden County" "Kent County" "Washington County" ...
 $ confirmed_cases      : int  111 3636 11548 5521 2496 3578 312 6255 527 307 ...
 $ deaths               : int  0 78 187 122 79 6 1 125 5 4 ...
 $ total_pop            : int  6203 160985 173145 126190 60383 83204 25191 151596 36825 30576 ...
 $ male_pop             : int  3135 78928 83544 61154 29705 40952 12504 76162 18214 15366 ...
 $ female_pop           : int  3068 82057 89601 65036 30678 42252 12687 75434 18611 15210 ...
 $ white_pop            : int  5929 143657 108627 115206 57523 71549 23895 129519 34245 29070 ...
 $ black_pop            : int  64 4091 41729 1621 285 2596 214 4425 337 272 ...
 $ asian_pop            : int  32 6144 3459 2436 579 1670 161 6690 711 186 ...
 $ hispanic_pop         : int  83 3542 11820 37

In [19]:
# Checking for missing values
sum(is.na(dataset_plus_census))

[1] 0

In [20]:
# Basic statistics
summary(dataset_plus_census)

    state           state_fips_code county_name        confirmed_cases    
 Length:3142        Min.   : 1.00   Length:3142        Min.   :      0.0  
 Class :character   1st Qu.:18.00   Class :character   1st Qu.:    796.2  
 Mode  :character   Median :29.00   Mode  :character   Median :   1916.5  
                    Mean   :30.28                      Mean   :   7558.9  
                    3rd Qu.:45.00                      3rd Qu.:   4955.0  
                    Max.   :56.00                      Max.   :1002614.0  
     deaths          total_pop           male_pop         female_pop     
 Min.   :    0.0   Min.   :      74   Min.   :     39   Min.   :     35  
 1st Qu.:   12.0   1st Qu.:   10945   1st Qu.:   5514   1st Qu.:   5460  
 Median :   31.0   Median :   25692   Median :  12798   Median :  12885  
 Mean   :  124.8   Mean   :  102166   Mean   :  50292   Mean   :  51873  
 3rd Qu.:   77.0   3rd Qu.:   67445   3rd Qu.:  33481   3rd Qu.:  34108  
 Max.   :13936.0   Max.   :1010

In [30]:
#check

# Convert state names to lowercase and count unique values
unique_state_names <- dataset_plus_census %>%
  mutate(state_name = tolower(state)) %>%  # Convert to lowercase
  summarize(unique_count = n_distinct(state_name))
print(unique_state_names)

# count of state FIPs codes
n_distinct(dataset_plus_census$state_fips_code)

# Convert county names to lowercase, remove "count" from the names, and count unique values
unique_county_names <- dataset_plus_census %>%
    mutate(county_name = tolower(county_name)) %>%
    mutate(count_name = gsub("county", "", county_name, ignore.case = TRUE)) %>%
    summarize(unique_count = n_distinct(county_name))
print(unique_county_names)

  unique_count
1           51


[1] 51

  unique_count
1         1878


There are 51 different states with their corresponding FIPs codes, and 1878 counties in the dataset.

In [34]:
# For heat map
install.packages("maps")

package 'maps' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\yamph\AppData\Local\Temp\RtmpWYQeHL\downloaded_packages


In [35]:
library(maps)


Attaching package: 'maps'


The following object is masked from 'package:purrr':

    map




In [36]:
# Heat map showing counties in Texas

# Get the county polygon data
counties_polygon <- as_tibble(map_data("county"))

# Filter for Texas counties and rename the subregion to county
counties_polygon_TX <- counties_polygon %>% dplyr::filter(region == "texas") %>% 
  rename(c(county = subregion)) 
counties_polygon_TX

long       lat      group order region county  
1    -95.75271  31.53560 2492  74520 texas  anderson
2    -95.76989  31.55852 2492  74521 texas  anderson
3    -95.76416  31.58143 2492  74522 texas  anderson
4    -95.72979  31.58143 2492  74523 texas  anderson
5    -95.74698  31.61008 2492  74524 texas  anderson
6    -95.72405  31.63873 2492  74525 texas  anderson
7    -95.75271  31.67311 2492  74526 texas  anderson
8    -95.76989  31.66738 2492  74527 texas  anderson
9    -95.77563  31.63300 2492  74528 texas  anderson
10   -95.79855  31.63873 2492  74529 texas  anderson
11   -95.81000  31.67311 2492  74530 texas  anderson
12   -95.79282  31.71321 2492  74531 texas  anderson
13   -95.82146  31.70748 2492  74532 texas  anderson
14   -95.87876  31.71321 2492  74533 texas  anderson
15   -95.87876  31.75332 2492  74534 texas  anderson
16   -95.91887  31.78197 2492  74535 texas  anderson
17   -95.94751  31.78197 2492  74536 texas  anderson
18   -95.98762  31.80489 2492  74537 texas  anderson
19   -95.98189  31.83927 2492  74538 texas  anderson
20   -95.99908  31.86218 2492  74539 texas  anderson
21   -95.98189  31.88510 2492  74540 texas  anderson
22   -96.02200  31.88510 2492  74541 texas  anderson
23   -96.03345  31.90229 2492  74542 texas  anderson
24   -96.01054  31.93667 2492  74543 texas  anderson
25   -96.02200  31.94813 2492  74544 texas  anderson
26   -96.07356  31.95959 2492  74545 texas  anderson
27   -96.06783  31.99396 2492  74546 texas  anderson
28   -96.06210  32.01688 2492  74547 texas  anderson
29   -95.44904  32.09710 2492  74548 texas  anderson
30   -95.44331  32.07418 2492  74549 texas  anderson
⋮    ⋮          ⋮        ⋮     ⋮     ⋮      ⋮       
4459  -98.94981 33.38625 2744  79230 texas  young   
4460  -98.42269 33.39198 2744  79231 texas  young   
4461  -99.44256 27.23841 2745  79233 texas  zapata  
4462  -99.39672 27.28425 2745  79234 texas  zapata  
4463  -99.11024 27.28425 2745  79235 texas  zapata  
4464  -98.96700 27.27279 2745  79236 texas  zapata  
4465  -98.96700 26.79151 2745  79237 texas  zapata  
4466  -99.03576 26.67692 2745  79238 texas  zapata  
4467  -99.15607 26.58524 2745  79239 texas  zapata  
4468  -99.17899 26.63108 2745  79240 texas  zapata  
4469  -99.20191 26.65973 2745  79241 texas  zapata  
4470  -99.20764 26.70556 2745  79242 texas  zapata  
4471  -99.27640 26.84880 2745  79243 texas  zapata  
4472  -99.31651 26.86599 2745  79244 texas  zapata  
4473  -99.34515 26.91183 2745  79245 texas  zapata  
4474  -99.37953 26.92329 2745  79246 texas  zapata  
4475  -99.39098 26.94048 2745  79247 texas  zapata  
4476  -99.40245 26.98058 2745  79248 texas  zapata  
4477  -99.42536 27.00350 2745  79249 texas  zapata  
4478  -99.45402 27.03215 2745  79250 texas  zapata  
4479  -99.46547 27.06080 2745  79251 texas  zapata  
4480  -99.44829 27.09517 2745  79252 texas  zapata  
4481  -99.43682 27.12382 2745  79253 texas  zapata  
4482  -99.44256 27.21549 2745  79254 texas  zapata  
4483  -99.44256 27.23841 2745  79255 texas  zapata  
4484  -99.40818 29.08334 2746  79257 texas  zavala  
4485  -99.40818 28.64216 2746  79258 texas  zavala  
4486 -100.11864 28.63643 2746  79259 texas  zavala  
4487 -100.11864 29.07188 2746  79260 texas  zavala  
4488  -99.40818 29.08334 2746  79261 texas  zavala

In [37]:
# Confirmed cases and deaths per 1000 for each county

#Summarize the data to get total confirmed cases and deaths by county
county_summary_1000 <- dataset_plus_census %>%
  group_by(county_name) %>%
  summarise(
    total_confirmed_cases = sum(confirmed_cases, na.rm = TRUE),
    total_deaths = sum(deaths, na.rm = TRUE),
    total_pop = sum(total_pop, na.rm = TRUE)
  ) %>%
  mutate(
    confirmed_cases_per_1000 = total_confirmed_cases / total_pop * 1000,
    deaths_per_1000 = total_deaths / total_pop * 1000
  )

In [41]:
# Do right join
counties_polygon_TX <- right_join(counties_polygon_TX, county_summary_1000, by = c("county" = "county_name"))

# plot the heat map
ggplot(counties_polygon_TX, aes(long, lat, group = group)) + 
  geom_polygon(aes(fill = confirmed_cases_per_1000)) +
  coord_fixed() +
    scale_fill_viridis_c() +
  labs(title = "Texas confirmed cases (per 1000) by county", 
fill = "Confirmed Cases per 1000") +
theme_minimal()

ERROR while rich displaying an object: Error in grid.Call.graphics(C_setviewport, vp, TRUE): non-finite location and/or size for viewport

Traceback:
1. sapply(x, f, simplify = simplify)
2. lapply(X = X, FUN = FUN, ...)
3. FUN(X[[i]], ...)
4. tryCatch(withCallingHandlers({
 .     if (!mime %in% names(repr::mime2repr)) 
 .         stop("No repr_* for mimetype ", mime, " in repr::mime2repr")
 .     rpr <- repr::mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
5. tryCatchList(expr, classes, parentenv, handlers)
6. tryCatchOne(expr, names, parentenv, handlers[[1L]])
7. doTryCatch(return(expr), name, parentenv, handler)
8. withCallingHandlers({
 .     if (!mime %in% names(repr::mime2repr)) 
 .         stop("No repr_* for mimetype ", mime, " in repr::mime2repr")
 .     rpr <- repr::mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content

## Chosen features for clustering

## Table with chosen features with their statistics

## Chosen measures for similarity/distance

## Cluster Analysis

### k-means Clustering

How a suitable number of clusters for this method was determined?

Compare the clusters

Visualization of evaluation of the clustering

Feature chosen as ground truth (label/target) for evaluation

### Hierarchical Clustering

### Third Clustering

### Fourth Clustering

## Questions to answer

What counties are similar in terms of makeup of the population, and the response to the virus?

What are the differences between the found groups of counties?

Can we identify groups that are more severely hit than others?